### PR TITLE
#161290908 Users Receive In-app and Email Notification

### PR DESCRIPTION
**Pull request by oluseyianani**
#### What does this PR do?
Enables users to receive in-app notifications and email notifications.
#### Description of Task to be completed?
  -  unit tests for email notification
  -  email notification using node mailer
  -  create migration of events and notification
  -  create events model and association
  -  create notification helpers for adding notification
  -  create route and controller for getting an in-app notification
  -  create an endpoint for notification opt-in and opt-out
  -  add notification worker to after article is published
  
#### How should this be manually tested?
- Create an account on the platform by following this instruction #11 
- Follow an Author on the platform by following this instruction #26 
- Following an Author allows a user to get notified when on the app or via Email when the Author posts an article
- If an Author, follow this instruction to post an article #13 
- Users can also get notifications on articles they like. To like an article, follow this instruction #18 

The route to get notifications (Read and Unread) - `/api/v1/notifications`
The route to opt-out or opt-in for notification - `/api/v1/users/notifications/opt/`

#### Any background context you want to provide?
A user must be logged in to access these routes

#### What are the relevant pivotal tracker stories?
#161290908

#### Screenshots (if appropriate)
screen shot 2018-11-26 at 3 45 43 pm

#### Questions:
N/A

### CODE REVIEW

**Review by ovieokeh — APPROVED** [review]
Please fix the conflicting files and the failing builds so that we can merge this.

File: server/src/helpers/following.js [hunk 1]
```diff
@@ -79,14 +79,14 @@ class FollowingHelper {
       if (!user) {
         return 'notAUser';
       }
-      const checkFollowing = await user.countFollower({ followerId });
```
Comment (ajimae) [inline server/src/helpers/following.js]: Nice work @anasey01 and @uwaelpis everything seems good, but am a little lost here?
Is this part of the notification feature?

File: server/src/helpers/following.js [hunk 2]
```diff
@@ -79,14 +79,14 @@ class FollowingHelper {
...
+      const checkFollowing = await user.hasFollower(followerId);
       const deleteFollowing = await user.removeFollower(followerId);
-      if (checkFollowing === 1 && deleteFollowing) {
```
Comment (ajimae) [inline server/src/helpers/following.js]: Also, from my understanding, this looks like a reactor, if yes, please specify it on the PR comment section so everyone will be aware of the changes made that are not directly related to the notification task. I suggest you put it under the **Background information section**

File: server/src/helpers/following.js [hunk 3]
```diff
@@ -79,14 +79,14 @@ class FollowingHelper {
...
+      if (checkFollowing && deleteFollowing) {
         user.action = 'unfollow';
         return user;
       }
       const add = await user.addFollower(followerId);
-      if (checkFollowing === 0 && add) {
```
Comment (ajimae) [inline server/src/helpers/following.js]: Here like wise

**Comment by ajimae** [comment]
Also, please fix these conflicting files below so the branch can be merged.

`
server/src/controllers/ArticleController.js
server/src/controllers/NotificationController.js
server/src/db/models/User.js
server/src/helpers/index.js
server/src/helpers/notification.js
server/src/helpers/templates.js
server/src/test/index.js
`

**Review by hope-uwa — APPROVED** [review]
Looks good

**Comment by ajimae** [comment]
Looks good now. Nice work @uwaelpis and @anasey01

**Review by ajimae — APPROVED** [review]
